### PR TITLE
Make extension callbacks part of the public MPS

### LINKED ISSUE #212: MPS does not support stack unwinding under Windows when it is managing code
New memory acquired by the arena when the arena grows is not registered with Windows for stack debugging purposes. This is an issue in cases where the MPS stores code objects on the heap. 

This can be solved by implementing Arena extension and contraction callback functions, corresponding or similar to those previously implemented for Configura, and for which we now have permission from Configura to implement as part of the core MPS. These are documented here: https://www.ravenbrook.com/project/mps/custom/cet/main/manual/html/custom/cet/extension.html

A use case for these callbacks would involve the mutator calling RtlInstallFunctionTableCallback when it is notified via callback that the arena has grown.

### PR DESCRIPTION
**Pull request by rptb1**
Fixes #212 .

Progress towards resolving #110 .

Part of the plan to meet Configura's requirements by separating and reviewing each implementation.

This branch cherry-picks and edits b928fa236178fb1bdbe20442c3f53b8e8a545a4b from #76 to only include the arena extension callback feature.

### CODE REVIEW

**Comment by rptb1** [comment]
So, VMChunkCreate (which should be named vmChunkCreate) is called from two places:

1. VMArenaCreate, where it creates the primary chunk, and
2. VMArenaGrow, where it creates all subsequent chunks.

We explicitly call the extended callback in both cases.

vmChunkDestroy (correctly named) is called from:

1. VMArenaDestroy, where it destroys all chunks via a TreeTraverse, and
2. vmChunkCompact, where we destroy chunks with nothing in them when trying to compact the arena.

We explicitly call the contracted callback only in vmChunkCompact.

It seems likely to me that we can call the extended and contracted callbacks from VMChunkCreate and vmChunkDestroy only, simplifying the code and eliminating the possibility of this inconsistency recurring.

**Comment by thejayps** [comment]
In the test extcon.c we observed on Rockhopper that the arena would mostly extend 50 times with the current test setup, but sometimes only 34 times. This seems odd and needs further investigation.

**Comment by thejayps** [comment]
Executing [proc.review.entry](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#entry)

1. Start time 15:45
2. proc.review.entry.criteria: It isn't clear how to select appropriate criteria for a review, other than the stipulation that entry.universal should always apply. 
!!!Procedure unclear: @thejayps continues anyway
I'm choosing to copy the entry criteria used for the recent [mps_pool_walk Pull Request](https://github.com/Ravenbrook/mps/pull/34) since the work is similar. There are also no other documents titled "entry" in the [Procedure Directory](https://github.com/Ravenbrook/mps/tree/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure). So I've chosen to continue by applying [entry.universal](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/entry.universal.rst), [entry.design](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/entry.design.rst), [entry.impl](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/entry.impl.rst).
3. entry.universal.author: This work was done by git cherry picking and manually isolating specific changes that form the implementation of extension callbacks from Configura's custom production MPS source. The recent work was performed by myself (I give permission) and @rptb1 (attempted to start proc.review.entry on this yesterday, but automated checks were failing. Permission assumed). Configura have given us permission to do this work: https://info.ravenbrook.com/mail/2022/01/18/12-53-38/0/
4. entry.universal.source-available: 2 mail threads: [The story behind "Ravenbrook MPS sources are out of sync with
Configura's sources"](https://info.ravenbrook.com/mail/2023/01/13/14-45-37/0/) and [Gathering information on Configura specific functionality](https://info.ravenbrook.com/mail/2023/03/23/16-08-02/0/) describe the approach to this work. (Links to the base email only, p4 grep can find all the subsequent responses).
5. entry.universal.source-approved - the current definition of this tag is unclear. 
!!!Procedure error: @thejayps continues anyway
I'll respond according to my interpretation: since many components of the source documents are not subject to procedures that have a "review exit", this criteria is difficult to apply. The intention of this criteria could be to guard against wasting time reviewing changes that themselves are based on as yet unreviewed changes which may fail their own review. The mail threads describing our work on this up to this point are not documents that are normally subject to any review process, and as such are UNREVIEWED.
6. entry.design.rfc this PR introduces extension callbacks, I do not know what historic review processes were in places when the original design documents were delivered, however I assume they were in place and took place. In this case, the implementation of callback functions is significantly trivial a concept that lacking reviewed design documentation should not pose a risk to the review. We must also remember that the aim of this work is to synchronise codebases, so the question "Is this the right design approach?" is not currently relevant. I'm choosing to continue.
7. Entry passed
8. End time 16:51
9. Entry took 66 minutes

**Comment by thejayps** [comment]
executing [proc.review.kickoff](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#53-review-kickoff)

[Link to rules directory for checkers](https://github.com/Ravenbrook/mps/tree/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure). All rules documents except rule.code.python.rst

**Comment by thejayps** [comment]
Here is [proc.review.check](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#54-review-checking) for checkers

**Comment by thejayps** [comment]
Roles:
@rptb1 is doing .consistency
@UNAA008 is doing .correctness of extcon.c
@thejayps is doing .source and .correctness of the underlying functionality (already existed in product document before change  linked it to public API)

**Comment by thejayps** [comment]
.ko.improve:
Record metrics and objectives:

2 objectives: Discover:
- How easy/hard is it for the review process to be assimilated by a new review leader
- How easy/hard is it for a non MPS familiar programmer to assess correctness of MPS code

**Comment by thejayps** [comment]
Review checking started at 1615. We'll check on the rate of checking at 1645.

File: manual/source/topic/arena.rst [hunk 1]
```diff
@@ -159,6 +159,17 @@ Client arenas
       may pause the :term:`client program` for. See
       :c:func:`mps_arena_pause_time_set` for details.
 
+    * :c:macro:`MPS_KEY_ARENA_EXTENDED` (type :c:type:`mps_fun_t`) is
+      a function that will be called when the arena is *extended*:
```
Comment (rptb1) [inline manual/source/topic/arena.rst:163]: m. Rather then "when" say "immediately after". rule.generic.clear

File: manual/source/topic/arena.rst [hunk 2]
```diff
@@ -159,6 +159,17 @@ Client arenas
...
+      that is, when it acquires a new chunk of address space from the
+      operating system. See :ref:`topic-arena-extension` for details.
+
+    * :c:macro:`MPS_KEY_ARENA_CONTRACTED` (type :c:type:`mps_fun_t`)
+      is a function that will be called when the arena is
+      *contracted*: that is, when it finishes with a chunk of address
```
Comment (rptb1) [inline manual/source/topic/arena.rst:169]: m. Rather than "when" say "immediately after". rule.generic.clear

File: code/mps.h [hunk 3]
```diff
@@ -120,6 +120,13 @@ typedef mps_addr_t (*mps_fmt_isfwd_t)(mps_addr_t);
 typedef void (*mps_fmt_pad_t)(mps_addr_t, size_t);
 typedef mps_addr_t (*mps_fmt_class_t)(mps_addr_t);
 
+/* Callbacks indicating that the arena has extended or contracted.
+ * These are used to register chunks with RtlInstallFunctionTableCallback
+ * <https://docs.microsoft.com/en-gb/windows/win32/api/winnt/nf-winnt-rtlinstallfunctiontablecallback>
+ * so that the client can unwind the stack through functions in the arena.
+ */
+typedef void (*mps_arena_extended_t)(mps_arena_t, mps_addr_t, size_t);
+typedef void (*mps_arena_contracted_t)(mps_arena_t, mps_addr_t, size_t);
```
Comment (rptb1) [inline code/mps.h:129]: m. Second argument to both of these functions is documented as `void *` not `mps_addr_t`. The latter is mean to indicate managed memory, which this arguably is not. rule.generic.self

File: manual/source/topic/arena.rst [hunk 4]
```diff
@@ -1040,3 +1051,79 @@ Arena introspection and debugging
...
+        /* ... other keyword arguments ... */
+        res = mps_arena_create_k(&arena, mps_arena_class_vm(), args);
+    } MPS_ARGS_END(args);
+
+The callback functions receive three arguments: ``arena`` (the arena
+being extended or contracted), ``base`` (the base address of the chunk
+of address space that has just been acquired from, or is about to be
+returned to, the operating system), and ``size`` (the size of the
+chunk, in bytes). They must not call any function in the MPS, and must
+not access any memory managed by the MPS.
```
Comment (rptb1) [inline manual/source/topic/arena.rst:1124]: m. Can these callbacks occur in signal / exception handlers?  We should mention it if so.  And check for other stuff (e.g. format methods) where we should mention that.  rule.generic.complete

File: code/extcon.c [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+  /* Clean up */
+  mps_root_destroy(testobj_root);
+  /* mps_root_destroy(stack_root);*/
+  mps_thread_dereg(thread);
+  mps_ap_destroy(obj_ap);
+  mps_pool_destroy(obj_pool);
+  mps_fmt_destroy(obj_fmt);
+  mps_arena_destroy(arena);
+#if 0
```
Comment (rptb1) [inline code/extcon.c:241]: m. Every #if'd out block must have an explanation.

**Review by rptb1 — COMMENTED** [review]
Executing [proc.review.check](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#54-review-checking)

1. Start time 16:15.
2. Checking consistency between https://www.ravenbrook.com/project/mps/custom/cet/main/manual/html/custom/cet/extension.html and updated manual.
3. Build the branch (`make clean test`) and the manual (`make -C manual html`) to check the results, check external links to Microsoft, etc.
4. Checked the changes to the MPS interface.
5. Checked the release notes include the above, and that they link correctly.
6. End time 16:45.  Somewhat limited by ME/CFS PEM.

**Review by thejayps — COMMENTED** [review]
M: Documentation doesn't clearly describe the behaviour of the callbacks. extended is called when the arena is created, however a new user may write their own program assuming that the callback is only called once an already existing arena is extended. We need to be clear that extension also means creation.

[Fixed: @thejayps added [a note](https://github.com/Ravenbrook/mps/blob/74d77c80f639f384f799506fc19c2dc15f09b94b/manual/source/topic/arena.rst?plain=1#L1127-L1145) in 74d77c80f

m: the contracted callback is not called when the arena is destoyed. This could be trivially important in practical use cases but is inconsistent with the above.

**Review by UNAA008 — COMMENTED** [review]
My review task was to check extcon.c.
I spent a total of 1 hour doing this.
Here are my comments:
m Copyright date in header inconsistent with that in footer.

q .limitation on not checking numbers of extensions and contractions.
   Is this a dangerous omission?

m An explanantion of why dylan interfaces are being used would be useful
   (line 28 is the first mention)

m Not all of the #if 0 exclusions have comments to explain them.
   I would prefer to see them marked so that it is obvious when they
   can be removed.

File: manual/source/topic/arena.rst [hunk 6]
```diff
@@ -1040,3 +1051,79 @@ Arena introspection and debugging
...
+-------------------------
+
+There are situations in which the :term:`client program` needs to be
+informed about the chunks of address space that an :term:`arena` is
+managing. To support this, the MPS allows the client program to
+specify two callback functions when creating a :term:`virtual memory
+arena`: one function is called when the arena is *extended* (that is,
+when it acquires a new chunk of address space from the operating
+system), and the other when the arena is *contracted* (that is, when
+it returns a chunk of address space to the operating system).
```
Comment (rptb1) [inline manual/source/topic/arena.rst:1074]: Nm. Are the contraction calls 1:1 with the extension calls?  Do we guarantee that?  I think we probably should.  rule.generic.complete.

**Review by rptb1 — COMMENTED** [review]
4 minor defects found during logging.

**Comment by thejayps** [comment]
Begin [logging](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#55-review-logging) at 1313

@UNAA008: 30mins + 30mins checking so far, found 3 minor issues and has a question.
@rptb1 5 minor issues, spent 30 mins yesterday checking
@thejayps 1 major issue and 1 minor issue, 30 mins checking yesterday.

@thejayps introduces major issue
.new.major @rptb1 the documentation for contraction also needs to be accurate and clear
@rptb1 talks through minor issues 
@UNAA008 talks through minor issues
Logging finished at 1406
Logging took 53 mins

**Comment by thejayps** [comment]
Begin [brainstorming](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#56-review-brainstorm)

started at 1425
brainstorm ideas by @rptb1 
1) Insist that documentation is produced alongside code, preferably by the same authors, and both are reviewed together.
2) In cases where the above hasn't happened - arrange for the author to check the documentation, and of course both checked against the original problem.
Action: ensure that rules exist to make these things happen. 
Brainstorming finished at 1439 - 14 mins

File: manual/source/topic/arena.rst [hunk 7]
```diff
@@ -1040,3 +1051,79 @@ Arena introspection and debugging
...
+being extended or contracted), ``base`` (the base address of the chunk
+of address space that has just been acquired from, or is about to be
+returned to, the operating system), and ``size`` (the size of the
+chunk, in bytes). They must not call any function in the MPS, and must
+not access any memory managed by the MPS.
+
+.. note::
+
+    Arena extension callbacks are only supported by :term:`virtual
+    memory arenas`.
```
Comment (thejayps) [inline manual/source/topic/arena.rst:1129]: @rptb1 this is possibly another case where https://github.com/Ravenbrook/mps/pull/223#discussion_r1197863623 applies,

**Comment by thejayps** [comment]
Executing [proc.review.edit](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#57-review-edit)

Start time 2023/05/23 0031

File: manual/source/topic/arena.rst (hunk 1 above)
Comment (thejayps) [reply manual/source/topic/arena.rst:163]: Changed to "immediately before" and "immediately after" as appropriate, with some other changes to words in the same paragraphs to conform

File: manual/source/topic/arena.rst (hunk 2 above)
Comment (thejayps) [reply manual/source/topic/arena.rst:169]: Fixed as per https://github.com/Ravenbrook/mps/pull/213#discussion_r1201298054

File: code/mps.h (hunk 3 above)
Comment (thejayps) [reply code/mps.h:129]: Fixed: changed to void *

File: code/extcon.c (hunk 5 above)
Comment (thejayps) [reply code/extcon.c:241]: Fix - explanation added in comments

File: manual/source/topic/arena.rst (hunk 4 above)
Comment (thejayps) [reply manual/source/topic/arena.rst:1124]: Raise #226

**Comment by thejayps** [comment]
> M: Documentation doesn't clearly describe the behaviour of the callbacks. extended is called when the arena is created, however a new user may write their own program assuming that the callback is only called once an already existing arena is extended. We need to be clear that extension also means creation.
> 
> m: the contracted callback is not called when the arena is destoyed. This could be trivially important in practical use cases but is inconsistent with the above.

Updated documentation in arena.rst to reflect current behaviour.

**Comment by thejayps** [comment]
Paused [proc.review.edit](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#57-review-edit)

Pause time 2023/05/23 0204

File: manual/source/topic/arena.rst (hunk 7 above)
Comment (rptb1) [reply manual/source/topic/arena.rst:1145]: In the case of https://github.com/Ravenbrook/mps/pull/223#discussion_r1197863623 the problem is that the information does not *help* the developer.  They can't necessarily tell whether an address is in an AMC or AMCZ pool, and what effect this had on the return code.  That's a consequence of the poor design of the interface and abuse of result codes. None of that applies in this case.  The fact stated is useful and complete.

File: code/extcon.c [hunk 8]
```diff
@@ -0,0 +1,295 @@
+/* extcon.c: ARENA EXTENDED AND CONTRACTED CALLBACK TEST
+ *
+ * $Id$
+ * Copyright (c) 2014-2022 Ravenbrook Limited.  See end of file for license.
```
[the remark this answers is not in the document]
Comment (rptb1) [reply code/extcon.c:4]: Ignore: duplicate issue already fixed in 2833d4505 .

File: code/extcon.c [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+#include "fmtdy.h"
+#include "fmtdytst.h"
+#include "mpsavm.h"
+#include "mpscamc.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+/* Number of test objects to allocate */
+#define N_TESTOBJ 100
+/* Number of dylan "slots" in each test object */
```
[the remark this answers is not in the document]
Comment (rptb1) [reply code/extcon.c:28]: Fix: Clarified in a20116a9a28fbd06e2772214d483287e571e9529

File: manual/source/topic/arena.rst (hunk 6 above)
Comment (rptb1) [reply manual/source/topic/arena.rst:1075]: Ignore: Already fixed in 74d77c80f639f384f799506fc19c2dc15f09b94b

**Comment by thejayps** [comment]
Executing [proc.review.exit] (https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst)

1. Start time 2102
2. All "conversations" according to git have been resolved.
3. Transgression: exit.universal.defect In the comment https://github.com/Ravenbrook/mps/pull/213#issuecomment-1544003201 @rptb1 raises a new major defect: "the documentation for contraction also needs to be accurate and clear" - there is no obvious response to this defect being raised in the comment.

Fixed: in fb93d8616d8b5db175037d201230c6bbc28ec067 where callbacks were redesigned to be more consistent, and documentation was updated.

**Comment by thejayps** [comment]
4. All questions have been answered
5. Transgression: the brainstorming ideas aren't linked to a defect: https://github.com/Ravenbrook/mps/pull/213#issuecomment-1544023867. I believe/recall that the suggestions were regarding the single major defect here: https://github.com/Ravenbrook/mps/pull/213#pullrequestreview-1420973870

.edit.act.raise : TBC

**Comment by thejayps** [comment]
Resuming .exit at 1030
5. exit.universal.record Start and end times are mostly recorded, and activity durations calculated in most of these cases. Timings for .edit need better recording. Transgression: no "estimated remaining defects" number is noted. Estimate here now:
6. .estimation: 2 Major defects were found. (one by @thejayps and a new one by @rptb1). Both were related, in fact the same defect applying to 2 opposite areas of code. We could argue that this is 1 defect, but go with 2 for safety. According to [.calc.defects-remaining](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#calc) we have 2 options: one calculation uses the number of pages - so we'll disregard this one until we know what a "page" means in github. Instead, use the "75%" method suggested by GavinM which I interpret to mean "We find ~75%" of major defects" -> 2 Major defects = 75% of the total. Total = 2 / 0.75 = 2.7 Major defects in total, so 0.7 (~1) defects remain.
7. .rates:  rates are slow and seemed appropriate while staff are trained (this was the first review of its kind). No danger that rates were too quick. Note that "the optimum rates" are not well defined in proc.exit.universal
8. .veto: no vetos (author and leader = @thejayps)
9. Revised change passed (exit.pass)

**Review by thejayps — APPROVED** [review]
[exit.pass](https://github.com/Ravenbrook/mps/pull/213#issuecomment-1587214689) at 1316

**Comment by thejayps** [comment]
Estimates (person mins)
.calc.manpower-used: (note task re [gender specific tags](https://github.com/Ravenbrook/mps/blob/branch/2023-01-19/review-procedure/procedure/review.rst#calc))
entry 66
plan 30
ko+check+log+bs (60+30+53+14)*3
edit 120 + 120*2
exit 85
~= 17 person-hours
.calc.manpower-saved:
2 Major defects *10 person hours = 20 hours

.calc.defects-remaining: 0-1 see calc [here](https://github.com/Ravenbrook/mps/pull/213#issuecomment-1587214689)

**Comment by thejayps** [comment]
Executing [proc.merge.pull-request](https://github.com/Ravenbrook/mps/blob/9dddbc5f9443b5dd8adfbce81bfe454237b9f3e2/procedure/pull-request-merge.rst) from #228 

1. Started 1410
2. Checklist passed (no tests were run locally)
3. End time 1420
4. Merge took 10 mins

**Pull request merged**